### PR TITLE
feat: V4 Quoter Integration

### PR DESCRIPTION
**Pull request by chefjackson**
---

## PR-Codex overview
This PR focuses on introducing support for `V4` liquidity pools in the `smart-router` package, enhancing the routing capabilities by adding new pool types and updating various components to accommodate the new `V4` features.

### Detailed summary
- Added export for `v4Pools` and `pool` types.
- Introduced `V4CL` and `V4BIN` pool types in various files.
- Implemented new functions for handling `V4` pools in routing and querying.
- Updated interfaces and types to include `V4` pools.
- Enhanced error handling and validations for `V4` pool operations.
- Modified tests to include scenarios for `V4` pools.
- Updated package dependencies to include `@pancakeswap/v4-sdk`.

> The following files were skipped due to too many changes: `packages/smart-router/evm/abis/ICLQuoter.ts`, `packages/smart-router/evm/abis/IBinQuoter.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Comment by chefjackson** [comment]
@SocketSecurity ignore npm/@sentry/node@5.30.0
@SocketSecurity ignore npm/@cspotcode/source-map-support@0.8.1
@SocketSecurity ignore npm/@emotion/is-prop-valid@0.8.8
@SocketSecurity ignore npm/@emotion/use-insertion-effect-with-fallbacks@1.0.1
@SocketSecurity ignore npm/@manypkg/find-root@1.1.0
@SocketSecurity ignore npm/@manypkg/get-packages@1.1.3
@SocketSecurity ignore npm/@changesets/types@4.1.0
@SocketSecurity ignore npm/@emotion/babel-plugin@11.11.0
@SocketSecurity ignore npm/@emotion/is-prop-valid@1.2.1
@SocketSecurity ignore npm/@cypress/xvfb@1.2.4
@SocketSecurity ignore npm/@material-ui/core@4.12.4
@SocketSecurity ignore npm/@material-ui/styles@4.11.5
@SocketSecurity ignore npm/@material-ui/system@4.12.2
@SocketSecurity ignore npm/@material-ui/utils@4.11.3
@SocketSecurity ignore npm/@nicolo-ribaudo/chokidar-2@2.1.8-no-fsevents.3
@SocketSecurity ignore npm/@cypress/request@2.88.12
@SocketSecurity ignore npm/@parcel/watcher-wasm@2.3.0
@SocketSecurity ignore npm/@microsoft/tsdoc-config@0.16.2
@SocketSecurity ignore npm/@nomicfoundation/solidity-analyzer@0.1.1
@SocketSecurity ignore npm/@json-rpc-tools/provider@1.7.6
@SocketSecurity ignore npm/@json-rpc-tools/utils@1.7.6
@SocketSecurity ignore npm/@json-rpc-tools/types@1.7.6
@SocketSecurity ignore npm/@babel/register@7.22.15
@SocketSecurity ignore npm/@lit-labs/ssr-dom-shim@1.1.2
@SocketSecurity ignore npm/@motionone/vue@10.16.4
@SocketSecurity ignore npm/@gnosis.pm/safe-react-gateway-sdk@3.5.2
@SocketSecurity ignore npm/@esbuild-plugins/node-globals-polyfill@0.1.1
@SocketSecurity ignore npm/@esbuild-plugins/node-modules-polyfill@0.1.4
@SocketSecurity ignore npm/@rushstack/node-core-library@3.61.0
@SocketSecurity ignore npm/@rushstack/rig-package@0.5.1
@SocketSecurity ignore npm/@humanwhocodes/config-array@0.11.13
@SocketSecurity ignore npm/@miniflare/cli-parser@2.13.0
@SocketSecurity ignore npm/@miniflare/core@2.13.0
@SocketSecurity ignore npm/@miniflare/d1@2.13.0
@SocketSecurity ignore npm/@miniflare/http-server@2.13.0
@SocketSecurity ignore npm/@miniflare/runner-vm@2.13.0
@SocketSecurity ignore npm/@miniflare/shared@2.13.0
@SocketSecurity ignore npm/@miniflare/storage-file@2.13.0
@SocketSecurity ignore npm/@miniflare/watcher@2.13.0
@SocketSecurity ignore npm/@rushstack/ts-command-line@4.17.1
@SocketSecurity ignore npm/@eslint/eslintrc@2.1.3
@SocketSecurity ignore npm/@storybook/components@7.5.3
@SocketSecurity ignore npm/@storybook/csf-plugin@7.5.3
@SocketSecurity ignore npm/@storybook/csf-tools@7.5.3
@SocketSecurity ignore npm/@storybook/node-logger@7.5.3
@SocketSecurity ignore npm/@storybook/core-common@7.5.3
@SocketSecurity ignore npm/@storybook/builder-manager@7.5.3
@SocketSecurity ignore npm/@storybook/builder-vite@7.5.3
@SocketSecurity ignore npm/@storybook/cli@7.5.3
@SocketSecurity ignore npm/@storybook/codemod@7.5.3

**Comment by chefjackson** [comment]
@SocketSecurity ignore npm/@storybook/core-server@7.5.3
@SocketSecurity ignore npm/@floating-ui/react-dom-interactions@0.12.0
@SocketSecurity ignore npm/@ledgerhq/iframe-provider@0.4.3
@SocketSecurity ignore npm/@babel/preset-env@7.23.3
@SocketSecurity ignore npm/@snapshot-labs/snapshot.js@0.4.110
@SocketSecurity ignore npm/@solana/web3.js@1.87.6
@SocketSecurity ignore npm/@pythnetwork/price-service-client@1.8.2
@SocketSecurity ignore npm/@changesets/read@0.6.0
@SocketSecurity ignore npm/@changesets/git@3.0.0
@SocketSecurity ignore npm/@changesets/apply-release-plan@7.0.0
@SocketSecurity ignore npm/@changesets/config@3.0.0
@SocketSecurity ignore npm/@changesets/pre@2.0.0
@SocketSecurity ignore npm/@changesets/write@0.3.0
@SocketSecurity ignore npm/@changesets/cli@2.27.1
@SocketSecurity ignore npm/@nomiclabs/hardhat-etherscan@3.1.8
@SocketSecurity ignore npm/@react-native/babel-preset@0.73.21
@SocketSecurity ignore npm/@react-native/metro-babel-transformer@0.73.15
@SocketSecurity ignore npm/@mui/base@5.0.0-beta.36
@SocketSecurity ignore npm/@mui/system@5.15.9
@SocketSecurity ignore npm/@mui/utils@5.15.9
@SocketSecurity ignore npm/@mui/private-theming@5.15.9
@SocketSecurity ignore npm/@mui/styled-engine@5.15.9
@SocketSecurity ignore npm/@mui/material@5.15.10
@SocketSecurity ignore npm/@react-native-community/cli@12.3.6
@SocketSecurity ignore npm/@react-native-community/cli-platform-android@12.3.6
@SocketSecurity ignore npm/@react-native-community/cli-platform-ios@12.3.6
@SocketSecurity ignore npm/@react-native-community/cli-clean@12.3.6
@SocketSecurity ignore npm/@react-native-community/cli-config@12.3.6
@SocketSecurity ignore npm/@react-native-community/cli-doctor@12.3.6
@SocketSecurity ignore npm/@react-native-community/cli-hermes@12.3.6
@SocketSecurity ignore npm/@react-native-community/cli-server-api@12.3.6
@SocketSecurity ignore npm/@react-native-community/cli-tools@12.3.6
@SocketSecurity ignore npm/@emotion/react@11.11.4
@SocketSecurity ignore npm/@emotion/is-prop-valid@1.2.2
@SocketSecurity ignore npm/@defi.org/web3-candies@4.20.17
@SocketSecurity ignore npm/@nomicfoundation/edr@0.3.3
@SocketSecurity ignore npm/@emotion/styled@11.11.5
@SocketSecurity ignore npm/@emotion/serialize@1.1.4
@SocketSecurity ignore npm/@next/eslint-plugin-next@14.2.3
@SocketSecurity ignore npm/@sentry/cli@2.32.1
@SocketSecurity ignore npm/@rollup/plugin-commonjs@26.0.1
@SocketSecurity ignore npm/@redocly/openapi-core@1.16.0
@SocketSecurity ignore npm/@opentelemetry/instrumentation@0.52.1
@SocketSecurity ignore npm/@opentelemetry/core@1.25.1
@SocketSecurity ignore npm/@opentelemetry/resources@1.25.1
@SocketSecurity ignore npm/@sentry/bundler-plugin-core@2.20.1
@SocketSecurity ignore npm/@prisma/instrumentation@5.16.1
@SocketSecurity ignore npm/@sentry/node@8.16.0
@SocketSecurity ignore npm/@sentry/nextjs@8.16.0
@SocketSecurity ignore npm/@sentry/vercel-edge@8.16.0

**Comment by chefjackson** [comment]
@SocketSecurity ignore npm/@orbs-network/twap-ui@0.11.2
@SocketSecurity ignore npm/@metamask/sdk-communication-layer@0.28.2

File: packages/smart-router/evm/abis/ICLPoolManager.ts [hunk 1]
```diff
@@ -0,0 +1,526 @@
+export const clPoolManagerAbi = [
```
Comment (ChefJerry) [inline packages/smart-router/evm/abis/ICLPoolManager.ts:1]: import from `v4-sdk`

File: packages/smart-router/evm/abis/ICLQuoter.ts [hunk 2]
```diff
@@ -0,0 +1,294 @@
+export const clQuoterAbi = [
```
Comment (ChefJerry) [inline packages/smart-router/evm/abis/ICLQuoter.ts:1]: import from `v4-sdk`

File: packages/smart-router/evm/constants/v4.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import { ChainId } from '@pancakeswap/chains'
+
+export const V4_CL_QUOTER_ADDRESSES = {
+  [ChainId.BSC_TESTNET]: '0xD2a534e28efb7e44Da94ae942F892E71F7462142',
```
Comment (ChefJerry) [inline packages/smart-router/evm/constants/v4.ts:4]: import from `v4-sdk`

File: packages/smart-router/evm/v3-router/utils/encodeV4ClRouteToPath.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        inputCurrency: outputCurrency,
+        path: [
+          ...path,
+          {
+            intermediateCurrency: getCurrencyAddress(outputCurrency),
+            fee: pool.fee,
+            hooks: pool.hooks ?? zeroAddress,
+            poolManager: pool.poolManager,
+            hookData: '0x',
+            parameters: encodeCLPoolParameters({
```
Comment (ChefJerry) [inline packages/smart-router/evm/v3-router/utils/encodeV4ClRouteToPath.ts:47]: missed `hooksRegistration`

**Comment by chefjackson** [comment]
@SocketSecurity ignore npm/@testing-library/react@12.1.5
@SocketSecurity ignore npm/@testing-library/dom@8.20.1
@SocketSecurity ignore npm/@testing-library/react@13.4.0
@SocketSecurity ignore npm/@reduxjs/toolkit@1.9.7
@SocketSecurity ignore npm/@tanstack/react-query@4.36.1
@SocketSecurity ignore npm/@tanstack/query-core@4.36.1
@SocketSecurity ignore npm/@tanstack/query-persist-client-core@4.36.1
@SocketSecurity ignore npm/@tanstack/react-query-devtools@4.36.1
@SocketSecurity ignore npm/@stylelint/postcss-css-in-js@0.37.3
@SocketSecurity ignore npm/@tinyhttp/app@1.3.0
@SocketSecurity ignore npm/@tinyhttp/etag@1.3.0
@SocketSecurity ignore npm/@tinyhttp/send@1.3.0
@SocketSecurity ignore npm/@storybook/theming@7.5.3
@SocketSecurity ignore npm/@storybook/router@7.5.3
@SocketSecurity ignore npm/@storybook/react-vite@7.5.3
@SocketSecurity ignore npm/@storybook/telemetry@7.5.3
@SocketSecurity ignore npm/@tanstack/query-core@5.52.0

File: packages/smart-router/evm/v3-router/utils/encodeV4ClRouteToPath.ts (hunk 4 above)
Comment (chefjackson) [reply packages/smart-router/evm/v3-router/utils/encodeV4ClRouteToPath.ts]: will add during integration

File: packages/smart-router/evm/constants/v4.ts (hunk 3 above)
Comment (chefjackson) [reply packages/smart-router/evm/constants/v4.ts:4]: will import during integration

**Pull request merged**